source/binding/Bitstream.cpp

```diff
@@ -622,13 +700,76 @@ static bool unpackConcatenation(const StreamingConcatenationExpression& lhs, Pac
             ASSERT(iterConcat == packed.cend() && !bit);
         }
         else {
-            auto rvalue = unpackBitstream(*stream->type, iter, iterEnd, bitOffset, dynamicSize);
+            auto& arrayType = *operand.type;
+            const Type* elemType = nullptr;
+            ConstantRange with;
+            if (stream->with) {
+                auto range = Bitstream::evaluateWith(arrayType, *stream->with, context);
+                if (!range)
+                    return false;
+                with = *range;
+                elemType = arrayType.getArrayElementType();
+                ASSERT(elemType != nullptr);
+                if (dynamicSize > 0 && !stream->with->width) {
+                    auto withSize = elemType->bitstreamWidth() * with.width();
+                    if (withSize >= dynamicSize)
+                        dynamicSize = 0;
+                    else
+                        dynamicSize -= withSize;
+                }
+            }
+            auto rvalue = unpackBitstream(
+                !stream->with
+                    ? arrayType
+                    : with.left == with.right
+                          ? *elemType
+                          : static_cast<const Type&>(FixedSizeUnpackedArrayType(*elemType, with)),
+                iter, iterEnd, bitOffset, dynamicSize);
             if (dryRun)
                 dryRun->emplace(std::move(rvalue));
             else {
-                LValue lvalue = stream->evalLValue(context);
+                LValue lvalue = operand.evalLValue(context);
                 if (!lvalue)
                     return false;
+
+                if (stream->with) {
+                    if (with.left != with.right && arrayType.isQueue() && !rvalue.isQueue())
+                        rvalue = constContainer(arrayType, rvalue.elements());
+
+                    if (!arrayType.hasFixedRange()) {
+                        ASSERT(with.left <= with.right);
+                        // When the array is a variable-size array, it shall be resized to
+                        // accommodate the range expression.
+                        const auto oldValue = lvalue.load();
+                        auto minSize = static_cast<uint32_t>(with.right + 1);
```

Comment:
Is with.right guaranteed to be positive here?

Reply:
Yes, "with" is the return value of Bitstream::evaluateWith function and at the end of Bitstream::evaluateWith, there are
```c++
    ASSERT(with.kind != WithRangeKind::Bit || result.left == result.right);
    ASSERT(result.left >= 0 && result.right >= 0);
    return result;
```